rust/rope/src/engine.rs

```diff
@@ -482,15 +471,222 @@ impl Engine {
+            let b_to_merge = &other.revs[base_index..];
+
+            let common = find_common(a_to_merge, b_to_merge);
+
+            let a_new = rearrange(a_to_merge, &common, self.deletes_from_union.len());
+            let b_new = rearrange(b_to_merge, &common, other.deletes_from_union.len());
+
+            let b_deltas = compute_deltas(&b_new[..], &other.text, &other.tombstones, &other.deletes_from_union);
+
+             rebase(a_new, b_deltas, self.text.clone(), self.tombstones.clone(), self.deletes_from_union.clone())
+        };
+
+        self.text = text;
+        self.tombstones = tombstones;
+        self.deletes_from_union = deletes_from_union;
+        self.revs.append(&mut new_revs);
+    }
+
+    /// Temporary hack until non-colliding ID generation is implemented
+    pub fn _set_rev_id_counter(&mut self, count: usize) {
+        self.rev_id_counter = count;
+    }
+}
+
+// ======== Generic helpers
+
+/// Move sections from text to tombstones and out of tombstones based on a new and old set of deletions
+fn shuffle_tombstones(text: &Rope, tombstones: &Rope,
+        old_deletes_from_union: &Subset, new_deletes_from_union: &Subset) -> Rope {
+    // Taking the complement of deletes_from_union leads to an interleaving valid for swapped text and tombstones,
+    // allowing us to use the same method to insert the text into the tombstones.
+    let inverse_tombstones_map = old_deletes_from_union.complement();
+    let move_delta = Delta::synthesize(text, &inverse_tombstones_map, &new_deletes_from_union.complement());
+    move_delta.apply(tombstones)
+}
+
+/// Move sections from text to tombstones and vice versa based on a new and old set of deletions.
+/// Returns a tuple of a new text `Rope` and a new `Tombstones` rope described by `new_deletes_from_union`.
+fn shuffle(text: &Rope, tombstones: &Rope,
+        old_deletes_from_union: &Subset, new_deletes_from_union: &Subset) -> (Rope,Rope) {
+    // Delta that deletes the right bits from the text
+    let del_delta = Delta::synthesize(tombstones, old_deletes_from_union, new_deletes_from_union);
+    let new_text = del_delta.apply(text);
+    // println!("shuffle: old={:?} new={:?} old_text={:?} new_text={:?} old_tombstones={:?}",
+    //     old_deletes_from_union, new_deletes_from_union, text, new_text, tombstones);
+    (new_text, shuffle_tombstones(text,tombstones,old_deletes_from_union,new_deletes_from_union))
+}
+
+// ======== Merge helpers
+
+/// Find an index before which everything is the same
+fn find_base_index(a: &[Revision], b: &[Revision]) -> usize {
+    assert!(a.len() > 0 && b.len() > 0);
+    assert!(a[0].rev_id == b[0].rev_id);
+    // TODO find the maximum base revision.
+    // this should have the same behavior, but worse performance
+    return 1;
+}
+
+/// Find a set of revisions common to both lists
+fn find_common(a: &[Revision], b: &[Revision]) -> BTreeSet<usize> {
+    // TODO: will the common revs always occur in the same order in both sets,
```

Comment:
I think the answer is probably no. If you have concurrent edits A and B, then it's pretty easy to imagine that one oplog will have (A, B'), while another has (B, A'), and that this merge will see both.